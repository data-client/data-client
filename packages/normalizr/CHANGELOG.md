# Change Log

## 0.10.0

### Minor Changes

- [#2912](https://github.com/reactive/data-client/pull/2912) [`922be79`](https://github.com/reactive/data-client/commit/922be79169a3eeea8e336eee519c165431ead474) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING CHANGE: `null` inputs are no longer filtered from Array or Object

  - `[]` and [schema.Array](https://dataclient.io/rest/api/Array) now behave in the same manner.
  - `null` values are now consistently handled everywhere (being retained).
    - These were already being retained in [nested Entities](https://dataclient.io/rest/guides/relational-data#nesting)
  - `undefined` is still filtered out.

### Patch Changes

- [`67f4e0b`](https://github.com/reactive/data-client/commit/67f4e0b45068da32d20e250267cb1cd2cea51226) Thanks [@ntucker](https://github.com/ntucker)! - Update README

- [`053e823`](https://github.com/reactive/data-client/commit/053e82377bd29f200cd7dfbc700da7a3ad7fa8d7) Thanks [@ntucker](https://github.com/ntucker)! - Update NextJS Demo link

## 0.9.5

### Patch Changes

- [`bb4b9583c5`](https://github.com/reactive/data-client/commit/bb4b9583c52e2b2fe45765af10b385b571901ee7) Thanks [@ntucker](https://github.com/ntucker)! - docs: Update readme

## 0.9.4

### Patch Changes

- [`d1b51af7ac`](https://github.com/reactive/data-client/commit/d1b51af7ac4a8a7c0559f478cc9503be8e61514c) Thanks [@ntucker](https://github.com/ntucker)! - Fix unpkg bundles by ensuring dependencies are built in order

## 0.9.3

### Patch Changes

- [#2818](https://github.com/reactive/data-client/pull/2818) [`fc0092883f`](https://github.com/reactive/data-client/commit/fc0092883f5af42a5d270250482b7f0ba9845e95) Thanks [@ntucker](https://github.com/ntucker)! - Fix unpkg bundles and update names

  - Client packages namespace into RDC
    - @data-client/react - RDC
    - @data-client/core - RDC.Core
    - @data-client/redux - RDC.Redux
  - Definition packages namespace top level
    - @data-client/rest - Rest
    - @data-client/graphql - GraphQL
    - @data-client/img - Img
    - @data-client/endpoint - Endpoint
  - Utility
    - @data-client/normalizr - normalizr
    - @data-client/use-enhanced-reducer - EnhancedReducer

## 0.9.2

### Patch Changes

- [`4ea0bc83f6`](https://github.com/reactive/data-client/commit/4ea0bc83f65f49cb2155f6aecdc5f8d1b168fd5e) Thanks [@ntucker](https://github.com/ntucker)! - Docs: Update repo links to reactive organization

## 0.8.0

### Minor Changes

- [#2784](https://github.com/reactive/data-client/pull/2784) [`c535f6c0ac`](https://github.com/reactive/data-client/commit/c535f6c0ac915b5242c1c7694308b7ee7aab16a1) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING CHANGES:

  - DELETE removed -> INVALIDATE
  - drop all support for legacy schemas
    - entity.expiresAt removed
    - Collections.infer does entity check
    - all Entity overrides for backcompat are removed - operates just like EntitySchema, except with extra validation

- [#2795](https://github.com/reactive/data-client/pull/2795) [`79e286109b`](https://github.com/reactive/data-client/commit/79e286109b5566f8e7acfdf0f44201263072d1d1) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING: [Schema Serializers](https://dataclient.io/rest/guides/network-transform#deserializing-fields) _must_ support function calls

  This means Date will no longer work like before. Possible migrations:

  ```ts
  class Ticker extends Entity {
    trade_id = 0;
    price = 0;
    time = Temporal.Instant.fromEpochSeconds(0);

    pk(): string {
      return `${this.trade_id}`;
    }
    static key = "Ticker";

    static schema = {
      price: Number,
      time: Temporal.Instant.from,
    };
  }
  ```

  or to continue using Date:

  ```ts
  class Ticker extends Entity {
    trade_id = 0;
    price = 0;
    time = Temporal.Instant.fromEpochSeconds(0);

    pk(): string {
      return `${this.trade_id}`;
    }
    static key = "Ticker";

    static schema = {
      price: Number,
      time: (iso: string) => new Date(iso),
    };
  }
  ```

- [#2792](https://github.com/reactive/data-client/pull/2792) [`35ccedceb5`](https://github.com/reactive/data-client/commit/35ccedceb53d91dd54dd996990c7c75719be2b85) Thanks [@ntucker](https://github.com/ntucker)! - BREAKING: Serializer schemas are only processed during denormalization

### Patch Changes

- [#2779](https://github.com/reactive/data-client/pull/2779) [`ff51e71f45`](https://github.com/reactive/data-client/commit/ff51e71f45857eb172f3fe05829e34c9abb68252) Thanks [@ntucker](https://github.com/ntucker)! - Update jsdocs references to dataclient.io

## 0.2.2

### Patch Changes

- 7b835f113a: Improve package tags

## 0.2.1

### Patch Changes

- 954e06e839: docs: Update README

## 0.2.0

### Minor Changes

- bf141cb5a5: Removed deprecated Endpoint.optimisticUpdate -> use Endpoint.getOptimisticResponse

### Patch Changes

- 87475a0cae: Mark expiresAt and useIncoming as deprecated in the public interface (still fine internally)
