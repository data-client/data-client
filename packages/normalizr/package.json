{
  "name": "@data-client/normalizr",
  "version": "0.9.5",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://dataclient.io/docs/concepts/normalization",
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/normalizr"
  },
  "keywords": [
    "normalize",
    "denormalize",
    "flux",
    "redux",
    "data",
    "model",
    "cache",
    "api",
    "react-native",
    "ios",
    "android",
    "web",
    "schema",
    "typescript",
    "validation",
    "relational",
    "reactive",
    "json"
  ],
  "files": [
    "dist/",
    "lib/",
    "node.mjs",
    "legacy",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "sideEffects": false,
  "scripts": {
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development yarn g:rollup",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production yarn g:rollup",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development yarn g:rollup && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:bundle": "yarn g:runs build:js:\\*",
    "build:clean": "yarn g:clean",
    "build:legacy-types": "yarn g:downtypes lib ts3.4",
    "prepublishOnly": "run build"
  },
  "author": "Nathaniel Tucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@types/node": "^20.0.0",
    "immutable": "4.2.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.17.0"
  }
}
