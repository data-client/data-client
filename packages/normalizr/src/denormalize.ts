import type { Schema, EntityInterface, UnvisitFunction } from './interface.js';
import { isEntity } from './isEntity.js';
import { denormalize as arrayDenormalize } from './schemas/Array.js';
import { isImmutable } from './schemas/ImmutableUtils.js';
import { denormalize as objectDenormalize } from './schemas/Object.js';
import type {
  Denormalize,
  DenormalizeNullable,
  DenormalizeCache,
  Path,
} from './types.js';
import WeakEntityMap, {
  type Dep,
  getEntities,
  type GetEntity,
  depToPaths,
} from './WeakEntityMap.js';

interface EntityCacheValue {
  dependencies: Dep[];
  value: [any, boolean, boolean];
}
const unvisitEntity = (
  entityOrId: Record<string, any> | string,
  schema: EntityInterface,
  unvisit: UnvisitFunction,
  getEntity: GetEntity,
  getCache: ReturnType<typeof getEntityCaches>,
  localCache: Record<string, Record<string, any>>,
  cycleCache: Record<string, Record<string, number>>,
  dependencies: Dep[],
  cycleIndex: { i: number },
): [denormalized: object | undefined, found: boolean, deleted: boolean] => {
  const entity =
    typeof entityOrId === 'object'
      ? entityOrId
      : getEntity({ pk: entityOrId, key: schema.key });
  if (
    typeof entity === 'symbol' &&
    (entity as symbol).toString().includes('DELETED')
  ) {
    return [undefined, true, true];
    // TODO: Change to this as breaking change once we only support newer entities
    // also remove `(entity as symbol).toString().includes('DELETED')` and do this for all symbols
    // return schema.denormalize(entity, unvisit);
  }

  if (typeof entity !== 'object' || entity === null) {
    return [entity as any, false, false];
  }

  const pk =
    // normalize must always place a string, because pk() return value is string | undefined
    // therefore no need to check for numbers
    typeof entityOrId === 'string'
      ? entityOrId
      : schema.pk(isImmutable(entity) ? (entity as any).toJS() : entity);
  // if we can't generate a working pk; this is hopeless so let's give them what's already there
  // otherwise, even when we aren't doing a lookup we want to turn the entityOrId object into the
  // expected class, and cache that class for referential equality. PK is used for global equality lookup.
  if (pk === undefined || pk === '' || pk === 'undefined') {
    return [entity, false, false];
  }

  const key = schema.key;
  if (!(key in localCache)) {
    localCache[key] = Object.create(null);
  }
  if (!(key in cycleCache)) {
    cycleCache[key] = Object.create(null);
  }
  const localCacheKey = localCache[key];
  const cycleCacheKey = cycleCache[key];

  let found = true;
  let deleted = false;

  // local cache lookup first
  if (!localCacheKey[pk]) {
    const globalCache: WeakEntityMap<object, EntityCacheValue> = getCache(
      pk,
      schema,
    );
    const [cacheValue] = globalCache.get(entity, getEntity);
    // TODO: what if this just returned the deps - then we don't need to store them

    if (cacheValue) {
      localCacheKey[pk] = cacheValue.value[0];
      // TODO: can we store the cache values instead of tracking *all* their sources?
      // this is only used for setting results cache correctly. if we got this far we will def need to set as we would have already tried getting it
      dependencies.push(...cacheValue.dependencies);
      return cacheValue.value;
    }
    // if we don't find in denormalize cache then do full denormalize
    else {
      const trackingIndex = dependencies.length;
      dependencies.push({ entity, path: { key, pk } });

      let entityCopy: any;
      if (schema.createIfValid) {
        entityCopy = localCacheKey[pk] = isImmutable(entity)
          ? schema.createIfValid(entity.toObject())
          : schema.createIfValid(entity);
        // TODO(breaking): remove once old verions no longer supported
      } else {
        entityCopy = entity;
        unvisit = withTrackedEntities(unvisit);
        unvisit.setLocal = entityCopy => (localCacheKey[pk] = entityCopy);
      }

      cycleCacheKey[pk] = trackingIndex;
      if (entityCopy === undefined) {
        // undefined indicates we should suspense (perhaps failed validation)
        found = false;
        deleted = true;
      } else {
        [localCacheKey[pk], found, deleted] = schema.denormalize(
          entityCopy,
          unvisit,
        );
      }
      delete cycleCacheKey[pk];

      // if in cycle, use the start of the cycle to track all deps
      // otherwise, we use our own trackingIndex
      const localKey = dependencies.slice(
        cycleIndex.i === -1 ? trackingIndex : cycleIndex.i,
      );
      const cacheValue: EntityCacheValue = {
        dependencies: localKey,
        value: [localCacheKey[pk], found, deleted],
      };
      globalCache.set(localKey, cacheValue);

      // start of cycle - reset cycle detection
      if (cycleIndex.i === trackingIndex) {
        cycleIndex.i = -1;
      }
    }
  } else {
    // cycle detected
    if (pk in cycleCacheKey) {
      cycleIndex.i = cycleCacheKey[pk];
    } else {
      // with no cycle, globalCacheEntry will have already been set
      dependencies.push({ entity, path: { key, pk } });
    }
  }

  return [localCacheKey[pk], found, deleted];
};

const getUnvisit = (
  entities: Record<string, Record<string, any>>,
  entityCache: DenormalizeCache['entities'],
  resultCache: DenormalizeCache['results'][string],
) => {
  const getEntity = getEntities(entities);
  const getCache = getEntityCaches(entityCache);
  const localCache: Record<string, Record<string, any>> = {};
  const dependencies: Dep[] = [];
  const cycleIndex = { i: -1 };
  const cycleCache = {};

  function unvisit(
    input: any,
    schema: any,
  ): [denormalized: any, found: boolean, deleted: boolean] {
    if (!schema) return [input, true, false];

    // null is considered intentional, thus always 'found' as true
    if (input === null) {
      return [input, true, false];
    }

    const hasDenormalize = typeof schema.denormalize === 'function';

    // deserialize fields (like Date)
    if (!hasDenormalize && typeof schema === 'function') {
      if (input instanceof schema) return [input, true, false];
      // field deserialization should never count against 'found' (whether to used inferred results)
      if (input === undefined) return [input, true, false];
      return [new schema(input), true, false];
    }

    if (input === undefined) {
      return [input, false, false];
    }

    if (!hasDenormalize && typeof schema === 'object') {
      const method = Array.isArray(schema)
        ? arrayDenormalize
        : objectDenormalize;
      return method(schema, input, unvisit);
    }

    if (isEntity(schema)) {
      return unvisitEntity(
        input,
        schema,
        unvisit,
        getEntity,
        getCache,
        localCache,
        cycleCache,
        dependencies,
        cycleIndex,
      );
    }

    if (hasDenormalize) {
      return schema.denormalize(input, unvisit);
    }

    return [input, true, false];
  }

  return (
    input: any,
    schema: any,
  ): [
    denormalized: any,
    found: boolean,
    deleted: boolean,
    entityPaths: Path[],
  ] => {
    // in the case where WeakMap cannot be used
    // this test ensures null is properly excluded from WeakMap
    const cachable = Object(input) === input && Object(schema) === schema;
    if (!cachable) {
      const ret = unvisit(input, schema);
      // this is faster than spread
      // https://www.measurethat.net/Benchmarks/Show/23636/0/spread-with-tuples
      return [ret[0], ret[1], ret[2], depToPaths(dependencies)];
    }

    let [ret, entityPaths] = resultCache.get(input, getEntity);

    if (ret === undefined) {
      ret = unvisit(input, schema);
      // we want to do this before we add our 'input' entry
      entityPaths = depToPaths(dependencies);
      // for the first entry, `path` is ignored so empty members is fine
      dependencies.unshift({ entity: input, path: { key: '', pk: '' } });
      resultCache.set(dependencies, ret);
    }

    return [ret[0], ret[1], ret[2], entityPaths as Path[]];
  };
};

const getEntityCaches = (entityCache: DenormalizeCache['entities']) => {
  return (pk: string, schema: EntityInterface) => {
    const key = schema.key;

    if (!(key in entityCache)) {
      entityCache[key] = Object.create(null);
    }
    const entityCacheKey = entityCache[key];
    if (!entityCacheKey[pk])
      entityCacheKey[pk] = new WeakMap<
        EntityInterface,
        WeakEntityMap<object, any>
      >();

    let wem: WeakEntityMap<object, any> = entityCacheKey[pk].get(schema) as any;
    if (!wem) {
      wem = new WeakEntityMap<object, any>();
      entityCacheKey[pk].set(schema, wem);
    }

    return wem;
  };
};

type DenormalizeReturn<S extends Schema> =
  | [
      denormalized: Denormalize<S>,
      found: true,
      deleted: false,
      entityPaths: Path[],
    ]
  | [
      denormalized: DenormalizeNullable<S>,
      found: boolean,
      deleted: true,
      entityPaths: Path[],
    ]
  | [
      denormalized: DenormalizeNullable<S>,
      found: false,
      deleted: boolean,
      entityPaths: Path[],
    ];

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const denormalize = <S extends Schema>(
  input: unknown,
  schema: S | undefined,
  entities: any,
  entityCache: DenormalizeCache['entities'] = {},
  resultCache: DenormalizeCache['results'][string] = new WeakEntityMap(),
): DenormalizeReturn<S> => {
  // undefined mean don't do anything
  if (schema === undefined) {
    return [input, true, false, []] as [any, boolean, boolean, any];
  }
  if (input === undefined) {
    return [undefined, false, false, []] as [any, boolean, boolean, any];
  }
  return getUnvisit(entities, entityCache, resultCache)(input, schema);
};

export const denormalizeSimple = <S extends Schema>(
  input: any,
  schema: S | undefined,
  entities: any,
  entityCache?: DenormalizeCache['entities'],
  resultCache?: DenormalizeCache['results'][string],
):
  | [denormalized: Denormalize<S>, found: true, deleted: false]
  | [denormalized: DenormalizeNullable<S>, found: boolean, deleted: true]
  | [denormalized: DenormalizeNullable<S>, found: false, deleted: boolean] =>
  denormalize(input, schema, entities, entityCache, resultCache).slice(
    0,
    3,
  ) as any;

// TODO(breaking): remove once unused
function withTrackedEntities(unvisit: UnvisitFunction): UnvisitFunction {
  // every time we nest, we want to unwrap back to the top.
  // this is due to only needed the next level of nested entities for lookup
  const originalUnvisit = unvisit.og || unvisit;
  const wrappedUnvisit = (input: any, schema: any) =>
    originalUnvisit(input, schema);
  wrappedUnvisit.og = unvisit;
  return wrappedUnvisit;
}
