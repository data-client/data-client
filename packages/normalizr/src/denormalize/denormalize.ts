import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import type { Denormalize, DenormalizeNullable } from '../types.js';
import { getEntities } from '../WeakEntityMap.js';

export const denormalize = <S extends Schema>(
  input: any,
  schema: S | undefined,
  entities: any,
): Denormalize<S> | DenormalizeNullable<S> => {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(getEntities(entities), new LocalCache())(input, schema)[0];
};
