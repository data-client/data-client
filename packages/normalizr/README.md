# normalizr

[![CircleCI](https://circleci.com/gh/data-client/data-client/tree/master.svg?style=shield)](https://circleci.com/gh/data-clientdata-clients)
[![Coverage Status](https://img.shields.io/codecov/c/gh/data-client/data-client/master.svg?style=flat-square)](https://app.codecov.io/gh/data-clientdata-clients?branch=master)
[![npm version](https://img.shields.io/npm/v/@data-client/normalizr.svg?style=flat-square)](https://www.npmjs.com/package/data-clients/normalizr) [![npm downloads](https://img.shields.io/npm/dmdata-clientks/normalizr.svg?style=flat-square)](https://www.npmjs.com/packagdata-clientoks/normalizr)

## Install

Install from the NPM repository using yarn or npm:

```shell
yarn add @data-client/normalizr
```

```shell
npm install --save @data-client/normalizr
```

## Motivation

Many APIs, public or not, return JSON data that has deeply nested objects. Using data in this kind of structure is often very difficult for JavaScript applications, especially those using [Flux](http://facebook.github.io/flux/) or [Redux](http://redux.js.org/).

## Solution

Normalizr is a small, but powerful utility for taking JSON with a schema definition and returning nested entities with their IDs, gathered in dictionaries.

## Documentation

- [Introduction](https://dataclient.io/docs/concepts/normalization)
- [Quick Start](https://dataclient.io/rest/api/schema)
- [API](https://dataclient.io/rest/api/Entity)
  - [normalize](./docs/api.md#normalizedata-schema)
  - [denormalize](./docs/api.md#denormalizeinput-schema-entities)
  - [schema](https://dataclient.io/rest/api/schema)

## Examples

- [Normalizing GitHub Issues](/examples/normalizr-github)
- [Relational Data](/examples/normalizr-relationships)
- [Interactive Redux](/examples/normalizr-redux)

## Quick Start

Consider a typical blog post. The API response for a single post might look something like this:

```json
{
  "id": "123",
  "author": {
    "id": "1",
    "name": "Paul"
  },
  "title": "My awesome blog post",
  "comments": [
    {
      "id": "324",
      "commenter": {
        "id": "2",
        "name": "Nicole"
      }
    }
  ]
}
```

We have two nested entity types within our `article`: `users` and `comments`. Using various `schema`, we can normalize all three entity types down:

```js
import { normalize, schema, Entity } from '@data-client/normalizr';

// Define a users schema
class User extends Entity {
  pk() {
    return this.id;
  }
}

// Define your comments schema
class Comment extends Entity {
  pk() {
    return this.id;
  }

  static schema = {
    commenter: User,
  };
}

// Define your article
class Article extends Entity {
  pk() {
    return this.id;
  }

  static schema = {
    author: User,
    comments: [Comment],
  };
}

const normalizedData = normalize(originalData, Article);
```

Now, `normalizedData` will be:

```js
{
  result: "123",
  entities: {
    "Article": {
      "123": {
        id: "123",
        author: "1",
        title: "My awesome blog post",
        comments: [ "324" ]
      }
    },
    "User": {
      "1": { "id": "1", "name": "Paul" },
      "2": { "id": "2", "name": "Nicole" }
    },
    "Comment": {
      "324": { id: "324", "commenter": "2" }
    }
  }
}
```

## Dependencies

None.

## Credits

Normalizr was originally created by [Dan Abramov](http://github.com/gaearon) and inspired by a conversation with [Jing Chen](https://twitter.com/jingc). Since v3, it was completely rewritten and maintained by [Paul Armstrong](https://twitter.com/paularmstrong).
Since v4, it was largely rewritten and maintained by [Nathaniel Tucker](https://twitter.com/npinp).
It has also received much help, enthusiasm, and contributions from [community members](https://github.com/ntucker/normalizr/graphs/contributors).
