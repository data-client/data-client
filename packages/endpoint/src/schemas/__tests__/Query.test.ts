// eslint-env jest
import { WeakEntityMap, inferResults } from '@data-client/normalizr';
import { IDEntity } from '__tests__/new';
import { fromJS } from 'immutable';

import { denormalizeSimple } from './denormalize';
import { schema, Query, Denormalize, DenormalizeNullable } from '../..';

let dateSpy: jest.SpyInstance<number, []>;
beforeAll(() => {
  dateSpy = jest
    // eslint-disable-next-line no-undef
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

const denormalize = denormalizeSimple;
class User extends IDEntity {
  name = '';
  isAdmin = false;
}

describe.each([
  ['direct', <T>(data: T) => data, <T>(data: T) => data],
  [
    'immutable',
    fromJS,
    (v: any) => (typeof v?.toJS === 'function' ? v.toJS() : v),
  ],
])(`input (%s)`, (_, createInput, createOutput) => {
  const SCHEMA_CASES = [
    ['All', new schema.Object({ results: new schema.All(User) })],
    [
      'Collection',
      new schema.Object({ results: new schema.Collection([User]) }),
    ],
  ] as const;
  if (_ === 'immutable') {
    delete (SCHEMA_CASES as any)[1];
  }

  describe.each(SCHEMA_CASES)(
    `${Query.name} denormalization (%s schema)`,
    (_, usersSchema) => {
      const sortedUsers = new Query(
        usersSchema,
        ({ results }, { asc } = { asc: false }) => {
          if (!results) return results;
          const sorted = [...results].sort((a, b) =>
            a.name.localeCompare(b.name),
          );
          if (asc) return sorted;
          return sorted.reverse();
        },
      );

      test('denormalize sorts', () => {
        const entities = {
          User: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
            3: { id: '3', name: 'Zeta' },
            4: { id: '4', name: 'Alpha' },
          },
          [new schema.Collection([User]).key]: {
            [new schema.Collection([User]).pk(undefined, undefined, '', [])]: [
              1, 2, 3, 4,
            ],
          },
        };
        const users: DenormalizeNullable<typeof sortedUsers.schema> | symbol =
          denormalize(
            inferResults(sortedUsers.schema, [], {}, entities),
            sortedUsers.schema,
            createInput(entities),
          );
        expect(users).not.toEqual(expect.any(Symbol));
        if (typeof users === 'symbol') return;
        expect(users && users[0].name).toBe('Zeta');
        expect(users).toMatchSnapshot();
      });

      test('denormalize sorts with arg', () => {
        const entities = {
          User: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
            3: { id: '3', name: 'Zeta' },
            4: { id: '4', name: 'Alpha' },
          },
          [new schema.Collection([User]).key]: {
            [new schema.Collection([User]).pk(undefined, undefined, '', [
              { asc: true },
            ])]: [1, 2, 3, 4],
          },
        };
        expect(
          denormalize(
            inferResults(sortedUsers.schema, [{ asc: true }], {}, entities),
            sortedUsers.schema,
            createInput(entities),
            {},
            new WeakEntityMap(),
            [{ asc: true }],
          ),
        ).toMatchSnapshot();
      });

      test('denormalizes should not be found when no entities are present', () => {
        const entities = {
          DOG: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        };
        const input = inferResults(sortedUsers.schema, [], {}, entities);

        const value = denormalize(
          createInput(input),
          sortedUsers.schema,
          createInput(entities),
        );

        expect(createOutput(value)).toEqual(undefined);
      });

      test('denormalize aggregates', () => {
        const userCountByAdmin = new Query(
          usersSchema,
          ({ results }, { isAdmin }: { isAdmin?: boolean } = {}) => {
            if (isAdmin === undefined) return results.length;
            return results.filter(user => user.isAdmin === isAdmin).length;
          },
        );
        const entities = {
          User: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake', isAdmin: true },
            3: { id: '3', name: 'Zeta' },
            4: { id: '4', name: 'Alpha' },
          },
          [new schema.Collection([User]).key]: {
            [new schema.Collection([User]).pk(undefined, undefined, '', [])]: [
              1, 2, 3, 4,
            ],
            [new schema.Collection([User]).pk(undefined, undefined, '', [
              { isAdmin: false },
            ])]: [1, 3, 4],
            [new schema.Collection([User]).pk(undefined, undefined, '', [
              { isAdmin: true },
            ])]: [2],
          },
        };
        const totalCount:
          | DenormalizeNullable<typeof userCountByAdmin.schema>
          | symbol = denormalize(
          inferResults(userCountByAdmin.schema, [], {}, entities),
          userCountByAdmin.schema,
          createInput(entities),
        );
        expect(totalCount).toBe(4);
        const nonAdminCount:
          | DenormalizeNullable<typeof userCountByAdmin.schema>
          | symbol = denormalize(
          inferResults(
            userCountByAdmin.schema,
            [{ isAdmin: false }],
            {},
            entities,
          ),
          userCountByAdmin.schema,
          createInput(entities),
          {},
          new WeakEntityMap(),
          [{ isAdmin: false }],
        );
        expect(nonAdminCount).toBe(3);
        const adminCount:
          | DenormalizeNullable<typeof userCountByAdmin.schema>
          | symbol = denormalize(
          inferResults(
            userCountByAdmin.schema,
            [{ isAdmin: true }],
            {},
            entities,
          ),
          userCountByAdmin.schema,
          createInput(entities),
          {},
          new WeakEntityMap(),
          [{ isAdmin: true }],
        );
        expect(adminCount).toBe(1);
        if (typeof totalCount === 'symbol') return;

        // typecheck
        totalCount !== undefined && totalCount + 5;
        // @ts-expect-error
        totalCount?.bob;
      });
    },
  );
});

describe('top level schema', () => {
  const sortedUsers = new Query(
    new schema.Collection([User]),
    (results, { asc } = { asc: false }) => {
      if (!results) return results;
      const sorted = [...results].sort((a, b) => a.name.localeCompare(b.name));
      if (asc) return sorted;
      return sorted.reverse();
    },
  );

  test('denormalize sorts', () => {
    const entities = {
      User: {
        1: { id: '1', name: 'Milo' },
        2: { id: '2', name: 'Jake' },
        3: { id: '3', name: 'Zeta' },
        4: { id: '4', name: 'Alpha' },
      },
      [new schema.Collection([User]).key]: {
        [new schema.Collection([User]).pk({}, undefined, '', [])]: [1, 2, 3, 4],
      },
    };
    const users: DenormalizeNullable<typeof sortedUsers.schema> | symbol =
      denormalize(
        inferResults(sortedUsers.schema, [], {}, entities),
        sortedUsers.schema,
        entities,
      );
    expect(users).not.toEqual(expect.any(Symbol));
    if (typeof users === 'symbol') return;
    expect(users && users[0].name).toBe('Zeta');
    expect(users).toMatchSnapshot();
  });

  test('denormalizes should not be found when no entities are present', () => {
    const entities = {
      DOG: {
        1: { id: '1', name: 'Milo' },
        2: { id: '2', name: 'Jake' },
      },
    };
    const input = inferResults(sortedUsers.schema, [], {}, entities);

    const value = denormalize(input, sortedUsers.schema, entities);

    expect(value).toEqual(undefined);
  });
});
