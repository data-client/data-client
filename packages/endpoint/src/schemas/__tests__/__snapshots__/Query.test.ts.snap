// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input (direct) Query denormalization (current) denormalize sorts 1`] = `
[
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
]
`;

exports[`input (direct) Query denormalization (current) denormalize sorts with arg 1`] = `
[
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
]
`;

exports[`input (direct) Query denormalization (legacy) denormalize sorts 1`] = `
[
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
]
`;

exports[`input (direct) Query denormalization (legacy) denormalize sorts with arg 1`] = `
[
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
]
`;

exports[`input (immutable) Query denormalization (current) denormalize sorts 1`] = `
[
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
]
`;

exports[`input (immutable) Query denormalization (current) denormalize sorts with arg 1`] = `
[
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
]
`;

exports[`input (immutable) Query denormalization (legacy) denormalize sorts 1`] = `
[
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
]
`;

exports[`input (immutable) Query denormalization (legacy) denormalize sorts with arg 1`] = `
[
  User {
    "id": "4",
    "isAdmin": false,
    "name": "Alpha",
  },
  User {
    "id": "2",
    "isAdmin": false,
    "name": "Jake",
  },
  User {
    "id": "1",
    "isAdmin": false,
    "name": "Milo",
  },
  User {
    "id": "3",
    "isAdmin": false,
    "name": "Zeta",
  },
]
`;
